### DISCUSSION TITLE
Both SSG/SSR variants of one page dependent on request headers

### DISCUSSION THREAD

**Discussion by fabb** [vercel/next.js#12100] [Help] [answered]
Ok, so this is my problem:
I have some simple pages (contact form, terms and conditions etc.) that I like to use SSG for. So far so great.
Then I have mobile apps that need to embed these pages in webviews. The problem is that I want to hide some elements like the header when shown in a webview in an app, because app user would get confused when they could navigate our whole website inside that webview (users wonder "Where is my tab bar gone?", "Why am I not logged in anymore?").

The simplest solution would of course be to not use SSG, but only SSR, and decide based on some custom request header whether to show the website header and other stuff. Not using SSG makes me sad though 😢.

Another idea would be to prepare 2 versions of my page, an SSG one and an SSR one and use an express middleware to route to the SSR version based on request headers. It's a bit boiler-platey though. Is there a better way to provide both an SSG and an SSR version of one page without duplicating the page file? Note that for further complication we also have a custom _app.tsx with `getInitialProps`.

**Comment by fabb** [comment]
For completion's sake: A solution using all-SSG would be to send some custom postMessage from the webview to my next app, but this would be more work on app side, and also the website header would shortly be visible until the app has finished hydration and only then disappear. Not nice. The apps could hide the webviews while loading, but that would make users stare at white areas even longer.

**Comment by fabb** [comment, marked as answer]
Ok what I did in the end was to duplicate all the affected pages, write a custom middleware that renders the corresponding duplicated page when a given request header was given, and in those SSG pages just re-export the original `getStaticProps` + wrap the original page in a HOC that sets a custom `Layout` on the page which I use to hide Headers/Footers. A bit complicated, and needs to duplicate all pages where I need this, but works.

**Reply by ilibilibom** [reply]
@fabb - Can you please post your solution here ? 
That would be very helpful 
Thanks 🙏

**Reply by fabb** [reply]
sure, no problem: https://gist.github.com/fabb/b1ca994abf2ded0145a917785bc796d7

note that when doing client routing, url paths need to be rewritten too conditionally, that's why i use a custom router to wrap the next.js router.
